04 - Reserving/Data Simulation/_scripts/feature_generation.py

```diff
@@ -0,0 +1,182 @@
+def feature_generation(V,lob,inflation,seed1):
+
+### Definition of the function Feature.Generation that generates features ###
+#############################################################################
+
+### Define the function Feature.Generation
+### We have the following inputs:
+### V = totally expected number of claims
+### LoB.dist = categorical distribution for the allocation of the claims to the 4 lines of business
+### inflation = growth parameters (per LoB) for the numbers of claims in the 12 accident years
+### seed1 = set the seed for reproducibility
+
+    import numpy as np
+    import pandas as pd
+    import scipy as sp
+    from scipy import stats
+    import math
+    import random
+
+# Weights in LoB.dist have to be nonnegative
+    if len([num for num in lob if num < 0]) > 0:
+        print("ERROR: The weights determining the distribution amongst the lines of business cannot be negative.")
+        features=pd.DataFrame(np.zeros((0,7)),columns=['ClNr', 'LoB', 'cc', 'AY', 'AQ', 'age', 'inj_part'])
+        return features
+
+# Preparation for the data set "features" 
+# Determine the number of claims per line of business
+    vdist = np.zeros((13,4))
+    random.seed(seed1)
+    rozkl=np.random.multinomial(V, lob, size=1)
+# Determine the number of claims per accident year (for all lines of business)
+    vdist[0]=rozkl
+    random.seed(seed1+1)
+    vmean=np.array([inflation,]*11)
+    vsd=np.array([np.abs(inflation),]*11)
+    vdist[2:13,:4]=np.random.normal(vmean,vsd)
+    vdist[2:13,:4]=np.cumsum(vdist[2:13,:4],axis=0)
+    vdist[1:13,:4]=np.exp(vdist[1:13,:4])
+    vdist_pom=vdist[1:13,:4].transpose()/vdist[1:13,:4].sum(axis=0)[:,None]
+    vdist[1:13,:4]=vdist_pom.transpose()
+    vdist_pom2=vdist[1:13,:4].transpose()*vdist[0,:4][:,None]
+    vdist[1:13,:4]=vdist_pom2.transpose()
+    random.seed(seed1+2)
+    vdist[1:13,:4]=np.random.poisson(vdist[1:13,:4],(12,4))
+    vdist[0,:4]=vdist[1:13,:4].sum(axis=0)[:,None].transpose()
+
+# Create the array where we will store the observations
+    features=pd.DataFrame(np.zeros((int(vdist[0,:4].sum()),7)),columns=['ClNr', 'LoB', 'cc', 'AY', 'AQ', 'age', 'inj_part'])
+
+# Store line of business
+    features.LoB=np.repeat(np.array([1,2,3,4]),vdist[0,:4].astype(int),axis=0).transpose()
```

Comment:
Comma

Reply:
From now on I am not commenting comment indentation and commas, the previous ones were just for indication